platforms/ide/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/DefaultBuildController.java

```diff
@@ -125,16 +125,26 @@ private ToolingModelBuilderResultInternal doGetModel(@Nullable Object target, To
         BuildTreeModelTarget scopedTarget = resolveTarget(target);
         try {
             Object model = controller.getModel(scopedTarget, modelRequestContext);
-            if (model instanceof ToolingModelBuilderResultInternal) {
-                return (ToolingModelBuilderResultInternal) model;
-            } else {
-                return ToolingModelBuilderResultInternal.of(model);
-            }
+            return getToolingModelBuilderResultInternal(modelRequestContext, model);
         } catch (UnknownModelException e) {
             throw (InternalUnsupportedModelException) new InternalUnsupportedModelException().initCause(e);
         }
     }
 
+    static ToolingModelBuilderResultInternal getToolingModelBuilderResultInternal(ToolingModelRequestContext modelRequestContext, @Nullable Object model) {
+        if (!(model instanceof ToolingModelBuilderResultInternal)) {
+            return ToolingModelBuilderResultInternal.of(model);
+        }
+        ToolingModelBuilderResultInternal resultInternal = (ToolingModelBuilderResultInternal) model;
+        if (modelRequestContext.inResilientContext()) {
+            return resultInternal;
+        }
+        if (resultInternal.getFailures().isEmpty()) {
+            return resultInternal;
+        }
+        throw resultInternal.throwOriginal();
```

Comment:
❓ This looks weird, can we avoid this magic, that tricks compiler?

We if we make `throwOriginal()` return `void` rather switch `if` around, e.g.:
```
        if (!resultInternal.getFailures().isEmpty()) {
            resultInternal.throwOriginal();
        }
        return resultInternal;
```